californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/SweepDeduplicator.java

```diff
@@ -23,101 +23,133 @@
 
-	/** The logger. */
-	private final static Logger LOGGER = Logger.getLogger(SweepDeduplicator.class.getCanonicalName());
-	
 	/** The hash map with all incoming messages. */
-	private ConcurrentHashMap<KeyMID, Exchange> incommingMessages;
-	
-	private NetworkConfig config;
+	private final Map<KeyMID, Exchange> incomingMessages = new ConcurrentHashMap<>();
+	private boolean running = false;
+	private ScheduledExecutorService scheduler;
 	private SweepAlgorithm algorithm;
-	
-	private ScheduledExecutorService executor;
-	
-	private boolean started = false;
-	
-	public SweepDeduplicator(NetworkConfig config) {
-		this.config = config;
-		incommingMessages = new ConcurrentHashMap<KeyMID, Exchange>();
-		algorithm = new SweepAlgorithm();
-	}
-	
-	public void start() {
-		started = true;
-		algorithm.schedule();
+
+	/**
+	 * Creates a new deduplicator from configuration values.
+	 * <p>
+	 * The following configuration values are used to initialize the
+	 * sweep algorithm used by this deduplicator:
+	 * <ul>
+	 * <li>{@link org.eclipse.californium.core.network.config.NetworkConfig.Keys#EXCHANGE_LIFETIME} -
+	 * an exchange is removed from this deduplicator if no messages have been received for this number
+	 * of milliseconds</li>
+	 * <li>{@link org.eclipse.californium.core.network.config.NetworkConfig.Keys#MARK_AND_SWEEP_INTERVAL} -
+	 * the interval at which to check for expired exchanges in milliseconds</li>
+	 * </ul>
+	 * 
+	 * @param config the configuration to use.
+	 */
+	public SweepDeduplicator(final NetworkConfig config) {
+		algorithm = new SweepAlgorithm(config);
 	}
-	
-	public void stop() {
-		started = false;
-		algorithm.cancel();
+
+	@Override
+	public synchronized void start() {
+		if (!running) {
+			if (scheduler == null || scheduler.isShutdown()) {
+				scheduler = Executors.newSingleThreadScheduledExecutor(new Utils.DaemonThreadFactory("Deduplicator"));
+			}
+			algorithm.schedule();
+			running = true;
+		}
 	}
-	
-	public void setExecutor(ScheduledExecutorService executor) {
-		stop();
-		this.executor = executor;
-		if (started)
-			start();
+
+	@Override
+	public synchronized void stop() {
+		if (running) {
+			algorithm.cancel();
+			scheduler.shutdown();
+			clear();
+			running = false;
+		}
 	}
-	
+
 	/**
 	 * If the message with the specified {@link KeyMID} has already arrived
 	 * before, this method returns the corresponding exchange. If this
 	 * KeyMID has not yet arrived, this method returns null, indicating that
 	 * the message with the KeyMID is not a duplicate. In this case, the
 	 * exchange is added to the deduplicator.
 	 */
-	public Exchange findPrevious(KeyMID key, Exchange exchange) {
-		Exchange previous = incommingMessages.putIfAbsent(key, exchange);
-		return previous;
+	@Override
+	public Exchange findPrevious(final KeyMID key, final Exchange exchange) {
+		return incomingMessages.putIfAbsent(key, exchange);
```

Comment:
Thanks for pointing this out, Simon :-) We *are* still targeting Java 7. I will change the type to `ConcurrentMap` ...